### ISSUE TITLE
[FR] Item subtypes default value for convenience methods

### ISSUE THREAD

**Issue by DerOetzi** [spacemanspiff2007/HABApp#256]
The openhab base_item get_value has a default parameter. The convenience methods for example (ContactItem.is_open) should have one as well.

And one step further it would be nice if there is a boolean flag for all of this get_value methods to update empty items with the default value. This is especially to set configuration items which are not linked to a channel on first read.

I have my own openhab scripting project using jython rules at the moment and will watch your project to find the right moment to migrate. You may have a look to my libraries directory to get some ideas: 

https://github.com/DerOetzi/openhab_creator/tree/master/openhab_creator/content/automation/libraries

Especially you may have a look at my item class:

https://github.com/DerOetzi/openhab_creator/blob/master/openhab_creator/content/automation/libraries/item.py

**Comment by spacemanspiff2007** [comment]
> The convenience methods for example (ContactItem.is_open) should have one as well.

How so? ``is_open`` will return false if the item state is NULL because obviously it is not open.

> This is especially to set configuration items which are not linked to a channel on first read.

I understand what you mean and I might think about it. However this is just a two liner which can be called if the rule is loaded:

```python
if item.value is None:
    item.oh_send_command('bla')
```
If you/we can come up with a very good name then maybe. :slightly_smiling_face:

**Comment by DerOetzi** [comment]
> > The convenience methods for example (ContactItem.is_open) should have one as well.
> 
> How so? `is_open` will return false if the item state is NULL because obviously it is not open.

Okay maybe ContactType.is_open was a bad example. Let's think about the logic of a configuration SwitchItem which can be NULL, ON or OFF. In my program logic sometimes I want to treat NULL same as OFF or as NULL by using SwitchItem.is_off(default_value=OFF)   

> 
> > This is especially to set configuration items which are not linked to a channel on first read.
> 
> I understand what you mean and I might think about it. However this is just a two liner which can be called if the rule is loaded:
> 
> ```python
> if item.value is None:
>     item.oh_send_command('bla')
> ```
> 
> If you/we can come up with a very good name then maybe. slightly_smiling_face

In my libraries item class has a method with signature Item.get_value(default_value=None,update_empty=False) which is the base for all subtype convenience method to make it to an oneliner :smile: 

```
if my_config_switch_item.is_off(OFF, True):

or

date = my_datetime_item.get_datetime(now(), True)
```

If update_empty is True an post_update is send to the item with the default_value to set correct  default value for the gui as well.

**Comment by spacemanspiff2007** [comment]
> In my program logic sometimes I want to treat NULL same as OFF or as NULL by using SwitchItem.is_off(default_value=OFF)

You can pick one of these:
```python
sw.is_on()        # true if sw is on
not sw.is_off()   # true if sw is on or Null
sw.get_value('ON') == 'ON' # true if sw is on or Null
```

> If update_empty is True an post_update is send to the item with the default_value to set correct default value for the gui as well.


Thank you for your input. Setting an item value with a function that will return the item value is imho confusing and counter intuitive and something I do not see me implementing.
You can however write your own library function (see docs on libs) and use it as you like.

**Comment by DerOetzi** [comment]
> 
> You can pick one of these:
> 
> ```python
> sw.is_on()        # true if sw is on
> not sw.is_off()   # true if sw is on or Null
> sw.get_value('ON') == 'ON' # true if sw is on or Null
> ```
> 

Those would be work. But both solutions would be not so nice in my opinion because line 2 with the `not` is "complicated" to understand and line 3 is okay to read but why "skipping" convenience method if following would be easy to implement

```python
sw.is_on('ON')
```

> > If update_empty is True an post_update is send to the item with the default_value to set correct default value for the gui as well.
> 
> Thank you for your input. Setting an item value with a function that will return the item value is imho confusing and counter intuitive and something I do not see me implementing. You can however write your own library function (see docs on libs) and use it as you like.

I thought about this again an have to thank you for your feedback indeed this is not very good coding practice. I think I will change this in my library as well. 

But I want to suggest another convenience method `is_empty` (name is just a suggestion) which encaspulate `item.value is None` as a named shorthand method.

**Comment by spacemanspiff2007** [comment]
implemented with 1.0.5: ``self.oh_post_update_if``
